addons/sale_stock/tests/test_sale_stock.py

```diff
@@ -723,3 +724,71 @@ def test_12_return_without_refund(self):
         return_picking.button_validate()
         # Checks the delivery amount (must still be 10).
         self.assertEqual(sale_order.order_line.qty_delivered, 10)
+
+    def test_13_sale_order_date_propagation(self):
+        """ Same test than test_04 but only with enough products in stock so that the reservation
+        is successful.
+        """
+
+        # Create 2 products with stock available
+        product_a = self.env['product.product'].create({'name': 'Des grosses Houppes', 'type': 'product'})
+        product_b = self.env['product.product'].create({'name': 'Des petites Houppes', 'type': 'product'})
+        wh = self.env['stock.warehouse'].search([('company_id', '=', self.env.company.id)], limit=1)
+        self.env['stock.quant']._update_available_quantity(product_a, wh.lot_stock_id, 2)
+        self.env['stock.quant']._update_available_quantity(product_b, wh.lot_stock_id, 2)
+
+        # Sell those 2 products
+        so = self.env['sale.order'].create({
+            'partner_id': self.partner.id,
+            'order_line': [
+                (0, 0, {'name': product_a.name, 'product_id': product_a.id, 'product_uom_qty': 1}),
+                (0, 0, {'name': product_b.name, 'product_id': product_b.id, 'product_uom_qty': 1}),
+            ],
+        })
+        so.action_confirm()
+
+        # Change the expected date of the sale order
+        new_date = fields.Datetime.now() + timedelta(days=2)
+        self.assertNotEquals(so.picking_ids.scheduled_date, new_date)
+        so.expected_date = new_date
+
+        # Check the date was propagated
+        self.assertEquals(len(so.picking_ids), 1)
+        self.assertEquals(so.picking_ids.scheduled_date, new_date)
+
+        # Deliver the picking
+        res_dict = so.picking_ids.button_validate()
+        immediate_transfer_wiz = self.env[(res_dict.get('res_model'))].browse(res_dict.get('res_id'))
+        immediate_transfer_wiz.process()
+        self.assertEquals(so.picking_ids.state, "done")
+        self.assertTrue(so.is_delivered)
+        picking_done = so.picking_ids[0]
+
+        # Update the two original sale order lines to create a second picking
+        so.write({
+            'order_line': [
+                (1, so.order_line.sorted()[0].id, {'product_uom_qty': 2}),
+                (1, so.order_line.sorted()[1].id, {'product_uom_qty': 2}),
+            ]
+        })
+        self.assertEquals(len(so.picking_ids), 2)
+        self.assertFalse(so.is_delivered)
+
+        # Change the expected date of the sale order
+        picking_to_do = so.picking_ids - picking_done
+        new_date_2 = fields.Datetime.now() + timedelta(days=4)
```

Comment:
complete the test with pick_ship and delay on rule and ensure everythings is correct for pick